src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs

```diff
@@ -77,61 +103,74 @@ public int Take()
                 return block.Array[index];
             }
 
+            return wasLastBlock ? -1 : TakeMultiBlock();
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private int TakeMultiBlock()
+        {
+            var block = _block;
             do
             {
-                if (wasLastBlock)
-                {
-                    return -1;
-                }
-                else
-                {
-                    block = block.Next;
-                    index = block.Start;
-                }
+                block = block.Next;
+                var index = block.Start;
 
-                wasLastBlock = block.Next == null;
+                // Always set wasLastBlock before checking .End to avoid race which may cause data loss 
+                var wasLastBlock = block.Next == null;
 
                 if (index < block.End)
                 {
                     _block = block;
                     _index = index + 1;
                     return block.Array[index];
                 }
+
+                if (wasLastBlock)
+                {
+                    return -1;
+                }
             } while (true);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Skip(int bytesToSkip)
         {
-            if (_block == null)
+            var block = _block;
+            if (block == null && bytesToSkip > 0)
             {
-                return;
+                ThrowInvalidOperationException_SkipMoreThanAvailable();
```

Comment:
**Note** Behaviour change, `Skip` now throws if `IsDefault` and `bytesToSkip > 0` rather than just returning; it didn't make sense it would throw if there weren't enough bytes, unless it was empty. Happy to revert if not correct.

/cc @cesarbs @halter73